src/main/kotlin/com/openlattice/organizations/ExternalDatabaseManagementService.kt

```diff
@@ -0,0 +1,369 @@
+                                "LEFT OUTER JOIN information_schema.table_constraints " +
+                                "on information_schema.constraint_column_usage.constraint_name = " +
+                                "information_schema.table_constraints.constraint_name " +
+                                "WHERE information_schema.columns.table_name = '$tableName' " +
+                                "$columnCondition"
+                )
+                while (rs.next()) {
+                    val columnName = rs.getString("column_name")
+                    val dataType = rs.getString("data_type")
+                    val position = rs.getInt("ordinal_position")
+                    val isPrimaryKey = rs.getString("constraint_type") == "PRIMARY KEY"
+                    val newColumn = OrganizationExternalDatabaseColumn(
+                            Optional.empty(),
+                            columnName,
+                            columnName,
+                            Optional.empty(),
+                            tableId,
+                            orgId,
+                            dataType,
+                            isPrimaryKey,
+                            position)
+                    newColumns.add(newColumn)
+                }
+            }
+        }
+        return newColumns
+    }
+
+    /*PRIVATE FUNCTIONS*/
+    private fun createPrivilegesSql(action: Action, privileges: List<String>, tableName: String, columnName: String, dbUser: String): String {
+        val privilegesAsString = privileges.joinToString(separator = ", ")
+        val tableNamePath = "$PUBLIC_SCHEMA.$tableName"
+        return if (action == Action.REMOVE) {
+            "REVOKE $privilegesAsString $columnName ON $tableNamePath FROM $dbUser"
+        } else {
+            "GRANT $privilegesAsString $columnName ON $tableNamePath TO $dbUser"
+        }
+    }
+
+    private fun getTableAndColumnNames(aclKey: AclKey): Pair<String, String> {
+        val securableObjectType = securableObjectTypes[aclKey]!!
+        val tableName: String
+        val columnName: String
+        val securableObjectId = aclKey.last()
+        if (securableObjectType == SecurableObjectType.OrganizationExternalDatabaseColumn) {
+            val organizationAtlasColumn = organizationExternalDatabaseColumns[securableObjectId]!!
+            tableName = organizationExternalDatabaseTables[organizationAtlasColumn.tableId]!!.name
+            columnName = organizationAtlasColumn.name
+        } else {
+            val organizationAtlasTable = organizationExternalDatabaseTables[securableObjectId]!!
+            tableName = organizationAtlasTable.name
+            columnName = ""
+        }
+        //add checks in here for map indexing??
+        return Pair(tableName, columnName)
+    }
+
+    private fun getDBUser(principalId: String): String {
+        val securePrincipal = securePrincipalsManager.getPrincipal(principalId)
```

Comment:
Either here or 1 call back ( in `addTrustedUser`) there should be a check, if the principals type is a user and if not, fail the call.

Reply:
Oh for sure ty